pkg/operator/ceph/reporting/reporting.go

```diff
@@ -33,25 +34,109 @@ import (
+
+// an object of a given type that has a nil reference is not the same as obj==nil (untyped nil)
+// (e.g., var cluster cephv1.CephCluster = nil ), so we must also check for nil via reflection
+func objIsNil(obj client.Object) bool {
+	return obj == nil || reflect.ValueOf(obj).IsNil()
+}
+
+// get the kind through the object API, but if that is empty, make a best guess via golang reflection
+func objKindOrBestGuess(obj client.Object) string {
+	// can't get any type info from an untyped nil
+	if obj == nil {
+		return unknownKind
+	}
+
+	if !reflect.ValueOf(obj).IsNil() {
+		kind := obj.GetObjectKind().GroupVersionKind().Kind
+		if kind != "" {
+			return kind
+		}
+	}
+
+	// typed nil, or object's typemeta is empty
+	t := reflect.TypeOf(obj)
+	if t.Kind() == reflect.Ptr {
+		return t.Elem().Name()
+	}
+	return t.Name()
+}
+
+func copyObject(obj client.Object) client.Object {
+	if obj == nil {
+		return nil // cannot copy nil object
+	}
+
+	if !reflect.ValueOf(obj).IsNil() {
+		return obj.DeepCopyObject().(client.Object) // deep copy the object if it's non-nil
+	}
+
+	// Otherwise, it is a nil object, but it has a type. We can use reflection to make an empty
+	// object to use as the copy in this case.
+	var nonNilCopy client.Object
+	t := reflect.TypeOf(obj)
+	if t.Kind() == reflect.Ptr {
+		innerType := t.Elem()
+		newObj := reflect.New(innerType)
+		nonNilCopy = newObj.Interface().(client.Object)
+	} else {
+		newObj := reflect.Zero(t)
+		nonNilCopy = newObj.Interface().(client.Object)
+	}
+
+	return nonNilCopy
+}
+
 // ReportReconcileResult will report the result of an object's reconcile in 2 ways:
 // 1. to the given logger
 // 2. as an event on the object (via the given event recorder)
-// The results of the object's reconcile should include the object, the reconcile response, and the
-// error returned by the reconcile.
+//
+// The results of the object's reconcile should include the object (NEVER NIL),
+// the reconcile response, and the error returned by the reconcile.
+//
+// The reconcile request is used to reference a given object that doesn't have a name.
+//
 // The function is designed to return the appropriate values needed for the controller-runtime
 // framework's Reconcile() method.
-func ReportReconcileResult(logger *capnslog.PackageLogger, recorder record.EventRecorder,
-	obj client.Object, reconcileResponse reconcile.Result, err error,
+func ReportReconcileResult(
+	logger *capnslog.PackageLogger,
+	recorder record.EventRecorder,
+	reconcileRequest reconcile.Request,
+	obj client.Object,
+	reconcileResponse reconcile.Result,
+	err error,
 ) (reconcile.Result, error) {
-	kind := obj.GetObjectKind().GroupVersionKind().Kind
-	nsName := fmt.Sprintf("%s/%s", obj.GetNamespace(), obj.GetName())
+	kind := objKindOrBestGuess(obj)
+
+	if objIsNil(obj) {
+		logger.Errorf("object associated with reconcile request %s %q should not be nil", kind, reconcileRequest)
+	}
+
+	objCopy := copyObject(obj)
```

Comment:
Why do we need to copy the object? In practice, the object had been passed by value before being passed by reference to this method, so it doesn't seem like there is any issue with changing fields in this object.

Reply:
For one, because we set namespace/name if the object is nameless, and we have no need to mutate the object in-place.

Reply:
Agreed on the approach if we're creating an API, but since rook owns both the callers and the callee, I just wonder if we can avoid the reflection calls. If the caller cares about the object mutation, what if they called `DeepCopy()` before passing it to `ReportReconcileResult()`?

Reply:
We may "own" both caller and callee, but that approach doesn't prevent developers from misusing the function. I am trying to limit the possible ways this can continue to break in the future for people misusing the caller, especially when the caller is a gigantic reconcile function where most people don't know that `ReportReconcileExists`.

Reply:
As discussed, let's go with this. My concern is really that reflection is a heavy hammer to swing, so I hope we don't have to use it very often.